macrosynergy/management/certificate_connection.py

```diff
@@ -0,0 +1,113 @@
+
+import base64
+import os
+import requests
+import json
+
+class DataQueryCert(object):
+    """
+    DataQuery REST web API class
+
+    For unittesting mock up methods
+
+        1. _create_api (return: mp_api)
+        2. _renew_token (mb_api fetch_token)
+        3. _fetch (mb_api get and post)
+
+    :param <str> username: username for login to REST API for
+        JP Morgan DataQuery.
+    :param <str> password: password.
+    :param <str> crt: string with location of public certificate.
+    :param <str> key: string with private key location.
+    """
+
+    def __init__(self, username: str,
+                 password: str,
+                 crt: str = "api_macrosynergy_com.crt",
+                 key: str = "api_macrosynergy_com.key"):
+
+        u_error = f"username must be a <str> and not <{type(username)}>."
+        assert isinstance(username, str), u_error
+
+        u_password = f"password must be a <str> and not <{type(password)}>."
+        assert isinstance(password, str), u_password
+        self.auth = base64.b64encode(bytes(f'{username:s}:{password:s}',
+                                           "utf-8")).decode('ascii')
+        self.headers = {"Authorization": f"Basic {self.auth:s}"}
+
+        self.base_url = "https://platform.jpmorgan.com/research/dataquery/api/v2"
+
+        # Key and Certificate.
+        self.key = self.valid_path(key, "key")
+
+        self.crt = self.valid_path(crt, "crt")
+        self.last_response = None
+
+    @staticmethod
+    def valid_path(directory: str, file_type: str):
+        """
+        Validates the key & certificate exist in the referenced directory.
+        """
+        error_type = file_type + " file must be a <str>."
+        assert isinstance(directory, str), error_type
+
+        try:
+            os.path.exists(directory)
+        except OSError as err:
+            print("OS error: {0}".format(err))
+        else:
+            return directory
+
+    def get_dq_api_result(self, url, params: dict = None):
+        """
+        Method used exclusively to request data from the API.
+        """
+        r = requests.get(url=url, cert=(self.crt, self.key),
+                         headers=self.headers, params=params)
+
+        return r
```

Comment:
Similar to above for OAuth, use context `with` statement, and also consolidate with `_fetch`, doesn't seem like we need both functions (unless we want fetch back in the `DataQueryInterface` itself and it is shared code between OAuth and Certificate Authentication).